keras_retinanet/utils/transform.py

```diff
@@ -0,0 +1,241 @@
+import numpy as np
+
+DEFAULT_PRNG = np.random
+
+
+def colvec(*args):
+    """ Create a numpy array representing a column vector. """
+    return np.array([args]).T
+
+
+def transform_aabb(transform, aabb):
+    """ Apply a transformation to an axis aligned bounding box.
+
+    The result is a new AABB in the same coordinate system as the original AABB.
+    The new AABB contains all corner points of the original AABB after applying the given transformation.
+
+    # Arguments
+        transform: The transormation to apply.
+        x1:        The minimum X value of the AABB.
+        y1:        The minimum y value of the AABB.
+        x2:        The maximum X value of the AABB.
+        y2:        The maximum y value of the AABB.
+    # Returns
+        The new AABB as tuple (x1, y1, x2, y2)
+    """
+    x1, y1, x2, y2 = aabb
+    # Transform all 4 corners of the AABB.
+    points = transform.dot([
+        [x1, x2, x1, x2],
+        [y1, y2, y2, y1],
+        [1,  1,  1,  1 ],
+    ])
+
+    # Extract the min and max corners again.
+    min_corner = points.min(axis=1)
+    max_corner = points.max(axis=1)
+
+    return [min_corner[0], min_corner[1], max_corner[0], max_corner[1]]
+
+
+def _random_vector(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random vector between min and max.
+    # Arguments
+        min: the minimum value for each component
+        max: the maximum value for each component
+    """
+    min = np.array(min)
+    max = np.array(max)
+    assert min.shape == max.shape
+    assert len(min.shape) == 1
+    return prng.uniform(min, max)
+
+
+def rotation(angle):
+    """ Construct a homogeneous 2D rotation matrix.
+    # Arguments
+        angle: the angle in radians
+    # Returns
+        the rotation matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [np.cos(angle), -np.sin(angle), 0],
+        [np.sin(angle),  np.cos(angle), 0],
+        [0, 0, 1]
+    ])
+
+
+def random_rotation(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random rotation between -max and max.
+    # Arguments
+        min:  a scalar for the minumum absolute angle in radians
+        max:  a scalar for the maximum absolute angle in radians
+        prng: the pseudo-random number generator to use.
+    # Returns
+        a homogeneous 3 by 3 rotation matrix
+    """
+    return rotation(prng.uniform(min, max))
+
+
+def translation(translation):
+    """ Construct a homogeneous 2D translation matrix.
+    # Arguments
+        translation: the translation 2D vector
+    # Returns
+        the translation matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [1, 0, translation[0]],
+        [0, 1, translation[1]],
+        [0, 0, 1]
+    ])
+
+
+def random_translation(min, max, prng = DEFAULT_PRNG):
```

Comment:
Do we want to distinguish between X/Y translation min/max?

Reply:
At this point I would say yes. These are generic transform functions. My initial reasoning was that you might want to scale the max and min with the image width and height. That isn't necessary now since the data generator does the scaling itself after generating the transform.

Still, I think it's a valid use case for a generic transform function.

Reply:
Damn you github, this comment thread is not outdated -.-

Reply:
But this code handles X and Y the same way right? Using the same min/max?

Reply:
no, min and max are 2D vectors (or a list/tuple/sequence of two elements)

Reply:
Ah I see, I missed that part then. :+1: